### ISSUE TITLE
cabal-head fails to build openapi3-3.2.4 "dist/build/example/example: copyFile: does not exist (No such file or directory)"

### ISSUE THREAD

**Issue by tchoutri** [haskell/cabal#9640]
**Describe the bug**

Compiling openapi3-3.2.4 with a "recent" `cabal-head` (installed through ghcup as said in the README) gives:

**To Reproduce**

```
$ ghcup install cabal --force -u https://github.com/haskell/cabal/releases/download/cabal-head/cabal-head-Linux-x86_64.tar.gz head
$ git clone https://github.com/flora-pm/flora-server
$ cd flora-server
$ cabal build
[…] # pour yourself a cup
Installing library in /home/tchoutri/.cabal/store/
  ghc-9.4.7/incoming/new-3485596/home/tchoutri/.cabal/
    store/ghc-9.4.7/openapi3-3.2.4-10ab5f1c59bbb45dbd67f02e0bbc49fff86fce693481f1e870150b3d6fa049d1/lib
Installing executable example in /home/tchoutri/.cabal/
  store/ghc-9.4.7/incoming/new-3485596/home/tchoutri/.cabal/store/ghc-9.4.7
    /openapi3-3.2.4-10ab5f1c59bbb45dbd67f02e0bbc49fff86fce693481f1e870150b3d6fa049d1/bin
Warning: The directory
/home/tchoutri/.cabal/store/ghc-9.4.7/incoming/new-3485596/home/tchoutri/
  .cabal/store/ghc-9.4.7/openapi3-3.2.4-10ab5f1c59bbb45dbd67f02e0bbc49fff86fce693481f1e870150b3d6fa049d1/bin
    is not in the system search path.
dist/build/example/example: copyFile: does not exist (No such file or directory)
Error: [Cabal-7125]
Failed to build openapi3-3.2.4 (which is required by test:flora-test from flora-1.0.14, exe:flora-server from flora-1.0.14 and others).
See the build log above for details.

```

**Expected behavior**
Not throw an error.

**System information**
 - Operating system: Fedora 38
 - `cabal`: cabal-head
 - `ghc`: 9.4.7

**Additional context**
I got my hands on a `cabal-head` from December 10th, that works well: [cabal-head-dec-10.tar.gz](https://github.com/haskell/cabal/files/14002901/cabal-head-dec-10.tar.gz).

Not too sure what to do with this information. :/

**Comment by Bodigrim** [comment]
> `dist/build/example/example: copyFile: does not exist (No such file or directory)`

`dist` instead of `dist-newstyle` is very suspicious here. Could anything go via `v1-*` code path?..

**Comment by geekosaur** [comment]
It also seems suspicious that an executable is being installed for a library dependency.

**Comment by geekosaur** [comment]
Never mind, building it locally I see `all, legacy fallback`. I also reproduce the install error. And also reproduce the weird-looking install paths (both `dist` and the duplication in the store path).

**Comment by geekosaur** [comment]
So for what it's worth, the store path shown is indeed invalid; `incoming` exists but contains only lock files, and in particular not `tmp-NNNNN`. The correct store path is the rest of the path, restarting from `/`.

"Legacy fallback" is because `openapi3-3.2.4` uses custom setup, but only for `cabal-doctests`. In particular it doesn't mess with the install.

**Comment by tchoutri** [comment]
@geekosaur thank you for reproducing, I feel less like a crazy person! :D  I put the blame automatically on the latest `cabal-head` because it's the only part of my toolchain that I regularly change. I was pretty lucky that @mau5mat hadn't updated his cabal-head since last December.

**Comment by TeofilC** [comment]
A colleague and I have also been able to reproduce this with `pretty-simple`, which also uses `cabal-doctests`

**Comment by Benjamin-McRae-Tracsis** [comment]
The build log emitted for this bit is attached.
[pretty-simple-4.1.2.0-d493aee511e73506dcc2cc921e2fdbf5d1ebc2e65ff6b5237e4a2cc57d544fcc.log](https://github.com/haskell/cabal/files/14009995/pretty-simple-4.1.2.0-d493aee511e73506dcc2cc921e2fdbf5d1ebc2e65ff6b5237e4a2cc57d544fcc.log)

**Comment by mpickering** [comment]
I have identified a good and bad commit, commencing bisecting.

**Comment by mpickering** [comment]
The bad commit is f70fc980dca50e7e62c5e803f56d2ef0c9fc7e18

**Comment by alt-romes** [comment]
Indeed, I'm on it. Fix incoming. Thanks @mpickering for marking this commit.

**Comment by alt-romes** [comment]
This is being surprisingly hard to fix.

A few notes on the comments in this thread:

* `dist` is used rather than `dist-newstyle` because the package is forced into `legacy fallback` mode because of the `Custom` setup.
* The path where things like haddocks are being written, for example `/home/tchoutri/.cabal/store/ghc-9.4.7/incoming/new-3485596/home/tchoutri/.cabal/store/ghc-9.4.7/openapi3-3.2.4-10ab5f1c59bbb45dbd67f02e0bbc49fff86fce693481f1e870150b3d6fa049d1/bin` is actually "right", at least in the sense that other cabal versions write here as well. This was not introduced by that commit.

By diffing the new patch against an older cabal, it seems the issue is that the following steps are omitted:
```
creating dist/build/example
creating dist/build/example/autogen
creating dist/build/example/autogen
creating dist/build/example/autogen
Preprocessing executable 'example' for openapi3-3.2.4..
Building executable 'example' for openapi3-3.2.4..
creating dist/build/example
creating dist/build/example/example-tmp
Running: /Users/romes/.ghcup/bin/ghc --make -no-link -fbuilding-cabal-package -O -static -outputdir dist/build/example/example-tmp -odir dist/build/example/example-tmp -hidir dist/build/example/example-tmp -stubdir dist/build/example/example-tmp -i -idist/build/example/example-tmp -iexamples -idist/build/example/autogen -idist/build/global-autogen -Idist/build/example/autogen -Idist/build/global-autogen -Idist/build/example/example-tmp -optP-include -optPdist/build/example/autogen/cabal_macros.h -hide-all-packages -Wmissing-home-modules -no-user-package-db -package-db /Users/romes/.cabal/store/ghc-9.8.1/package.db -package-db dist/package.conf.inplace -package-id sn-2.2.1.0-f68a9a78 -package-id base-4.19.0.0-e537 -package-id lns-5.2.3-4d20fb14 -package-id pnp3-3.2.4-9520c7ce -package-id text-2.1-0f57 -XHaskell2010 examples/hackage.hs -hide-all-packages
[1 of 1] Compiling Main             ( examples/hackage.hs, dist/build/example/example-tmp/Main.o )
Linking...
Running: /Users/romes/.ghcup/bin/ghc --make -fbuilding-cabal-package -O -static -outputdir dist/build/example/example-tmp -odir dist/build/example/example-tmp -hidir dist/build/example/example-tmp -stubdir dist/build/example/example-tmp -i -idist/build/example/example-tmp -iexamples -idist/build/example/autogen -idist/build/global-autogen -Idist/build/example/autogen -Idist/build/global-autogen -Idist/build/example/example-tmp -optP-include -optPdist/build/example/autogen/cabal_macros.h -hide-all-packages -Wmissing-home-modules -no-user-package-db -package-db /Users/romes/.cabal/store/ghc-9.8.1/package.db -package-db dist/package.conf.inplace -package-id sn-2.2.1.0-f68a9a78 -package-id base-4.19.0.0-e537 -package-id lns-5.2.3-4d20fb14 -package-id pnp3-3.2.4-9520c7ce -package-id text-2.1-0f57 -XHaskell2010 examples/hackage.hs -o dist/build/example/example -hide-all-packages
[2 of 2] Linking dist/build/example/example
ld: warning: ignoring duplicate libraries: '-lm'
```
That is, specifically, the directory `dist/build/example` is not created, and the executable is not built. However, in the install step, we try to copy it over.

Now, why is the executable not being built? It's what I'm trying to figure out -- I don't know why answering that question is being so troublesome -- it may be too late in the day, and perhaps I should unfortunately pick this up tomorrow.

**Comment by alt-romes** [comment]
To reinforce my hypothesis:

A good cabal prints
```
Component build order: library, executable 'example'
```

whereas a bad one just says
```
Component build order: library
```

**Comment by alt-romes** [comment]
Fix in !9650

**Comment by Mikolaj** [comment]
For the record, the PR with the faulty commit is https://github.com/haskell/cabal/pull/9524

**Comment by alt-romes** [comment]
To summarize, in #9524 we started passing build arguments such as `lib:openapi3` to `./Setup build` (called for building legacy-fallbak packages, e.g. ones that use `Custom`).

This has the benefit of guaranteeing that *unneeded targets* in such legacy fallback packages, like `exe:example`, *are not built*.

However, it missed the *copy args*. The discrepancy between passing build args but not copy args means we will only build some targets, but try to copy all of them (some of which won't have been built).

The patch I put up fixes it by updating the copy args. Now, we get the improvement without the bug.

I've also minimized the issue and created a test.
